### PR TITLE
Add LEP for Live Upgrade For Data Engine of V2 Volumes

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue Longhorn/longhorn#6001

### CODE REVIEW

**Comment by derekbit** [comment]
@DamiaSan @shuo-wu @innobead Could you take a look at the data plane design? Thank you.

**Comment by derekbit** [comment]
@DamiaSan @shuo-wu @innobead Can you take a look at the data plane design? Thank you.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+      ```
+    - EngineCreateRequest
+      During an upgrade, the initiator and target reside on different nodes. `initiator_address` and `target_address` specify the respective addresses of the initiator and target.
+      ```
+      message EngineCreateRequest {
+        ...
+        bool upgrade_required = 7;
+        string initiator_address = 8;
+	    string target_address = 9;
+      }
```
Comment (DamiaSan) [inline enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:117]: Don't we need here also target and initiator ports?

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      ```
+    - InstanceDeleteTargetRequest
+      ```
+      message InstanceDeleteTargetRequest {
+      	DataEngine data_engine = 1;
+      	string name = 2;
+      	string type = 3;
+      }
+      ```
```
Comment (DamiaSan) [inline enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:204]: Could you make a brief description of the fields of these last 4 messages?

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  - Replica Rebuilding
+      - Rebuilding failed replicas
+
+  Then, the upgrade process is complete once all volumes on the upgrading node are healthy.
+
+
+  Here is an example for upgrading instance-manager for v2 engine on the upgrading node `node1`:
+
+  ![An example for upgrading instance-manager for v2 engine on the upgrading node `node1`](image/v2-volume-upgrade.png)
+
```
Comment (DamiaSan) [inline enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:86]: Is the title of the fifth figure correct? It is the same of the previous one, "Switch over target to ..."

**Review by DamiaSan — COMMENTED** [review]
Just a few questions, but the general concept is correct for me.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md (hunk 1 above)
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:117]: Yes, we need. address is the join of an ip and a host.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md (hunk 3 above)
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:86]: Good catch! Updated. Thank you.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md (hunk 2 above)
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:208]: Added. Thank you.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      - Resume the linear device mapper and continue IO processing.
+        ```
+        InstanceResume()
+        |-> EngineResume()
+        ```
+  - Instance Manager Upgrade
+      - If the existing instance-manager pod does not have any running engines with targets, the instance-manager and its pod will be deleted by the node controller.
+      - Replicas managed by the deleted instance-manager are marked as ERROR, causing any volume with replicas on the upgrading node to become degraded.
+      - A new instance-manager is then created and starts running.
+  - Switch Over Target Back
```
Comment (shuo-wu) [inline enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:71]: Similarly, will Longhorn revert to the temporary target if there is something wrong with the new target?

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+## Motivation
+
+### Goals
+
+- Support both destructive and live upgrades of the data engine for a v2 volume with multiple replicas.
+- Upgrade volumes one node at a time.
+
+### Non-goals [optional]
+
+- Support live upgrades of the data engine for a v2 volume with a single replica.
```
[the remark this answers is not in the document]
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:20]: No. It is a hard limit.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        ```
+        InstanceSwitchOverTarget()
+        |-> EngineSwitchOverTarget()
+        ```
+      - Delete the old target for the volume on the upgrading node.
+        ```
+        InstanceDeleteTarget()
+        |-> EngineDeleteTarget()
+        ```
+      - Resume the linear device mapper and continue IO processing.
```
[the remark this answers is not in the document]
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:62]: > due to this case, it's better to resume the dm device IO before deleting the old target.

Yes, but I found the nvme will somehow crash if deleting the old target after resuming IO. I'm still investigating the root cause.

[the remark this answers is not in the document]
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:62]: > due to this case, it's better to resume the dm device IO before deleting the old target.

Ideally, yes. But we need to overcome https://github.com/longhorn/longhorn/pull/8814#discussion_r1657944110 first.

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md (hunk 4 above)
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md:71]: Same issue for now https://github.com/longhorn/longhorn/pull/8814#discussion_r1657944110

File: enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+### User Stories
+
+After upgrading the Longhorn system, users may want to upgrade the data engine of active v2 volumes to access new features or bug fixes without service interruption.
+
+### API changes
+
+## Design
+
+Unlike v1 volumes, upgrading the data engine for individual v2 volumes is not possible. This limitation exists because each node disk can only be managed by a single SPDK target daemon (spdk_tgt) within an instance-manager pod at any given time. The engine and replicas of a volume cannot be moved to a newly created instance-manager pod using a new (default) instance-manager image. This means the existing instance-manager pod must be removed before deploying a new one with the new image.
```
[the remark this answers is not in the document]
Comment (derekbit) [reply enhancements/20240624-live-upgrade-for-data-engine-of-v2-volumes.md]: Addressed. Thank you.

**Comment by derekbit** [comment]
Replaced by https://github.com/longhorn/longhorn/pull/9807

**Pull request closed without merging**